Feature/random provider

junniest opened the pull request:
Implemented `RandomSource` to use in the crate and added a fake random to use in testing.

junniest committed:
Initial version of the global random

Initial version of the global rng for the whole crate, currently using StdRng
which is cryptographically secure but not necessarily reproducible because it depends
on the system configuration and might change with library versions.

junniest committed:
Generalisation to work with different RNG backends

Generalised the global random implementation to allow different RNG backends.
The crate-wide RNG still uses StdRng.

junniest committed:
Examples for phylo::random use

junniest committed:
Add trait for RandomSource

junniest committed:
Inject rng into topo optimiser for testing

Injecting a rng provider `RandomSource` into topology optimiser to
allow using a fake deterministic rng in tests and remove the `deterministic`
feature.

junniest committed:
FakeGenerator implementation for testing

FakeGenerator is a deterministic struct that implements `RandomSource` for testing

junniest committed:
Inject rng provider into PhyloInfoBuilder

Inject rng provider into PhyloInfoBuilder to pass to the NJ tree builder

junniest committed:
Injecting rng provider into tree building

junniest committed:
RNG usage examples wo global static random

junniest committed:
Add `ntimestamp` to seed rng

junniest committed:
Simplifying FakeGenerator

junniest committed:
FakeGenerator tests

junniest committed:
Add more types to FakeGenerator::gen()

junniest committed:
Fix FakeGenerator::get_probability()

junniest committed:
Fix test names and small refactor

junniest committed:
Remove redundant test

junniest committed:
Remove gen_range from RandomSource

Remove gen_range from RandomSource trait but keep it as an impl for RandomGenerator.

junniest committed:
Fix fake rng test

junniest committed:
Refactor test names

junniest committed:
Remove unnecessary `deterministic` feature

junniest committed:
Use fake rng in benches for determinism

junniest committed:
Remove `deterministic` feature from CI

junniest committed:
Immediate panic for unsupported types in fakerng

kalxed commented on phylo/examples/rng_examples.rs:
```diff
@@ -0,0 +1,79 @@
+use rand::rngs::{SmallRng, StdRng};
```
I still have to see the extent to which these are used but just wanted to highlight they are non-portable meaning if any test result uses SmallRng even though it's a PRNG it might have different results on different platforms or in future versions

junniest replied:
These are just there as an example, `SmallRng` is only used to show that it's possible to create a `RandomSource<SmallRng>`. Both `SmallRng` and `StdRng` are potentially non-portable from system to system, but the results for running anything with them should be still deterministic if the setup is fixed (i.e. the code is run on the same machine, same system version, same seed). Although I was planning to replace `StdRng` as the default generator and got stuck due to a dependency clash, thanks for reminding me to look into that again!

kalxed commented on phylo/examples/rng_examples.rs:
I assume this file is simply for testing RNG examples though not automated? I think it makes more sense so people can see examples of random usage.

junniest replied:
This is actually not for testing, it's an example binary that you can run with `cargo run --example rng_examples`. I didn't know cargo could do that until very recently, but it's a nice way to provide examples other than the doctests.

kalxed commented on phylo/src/optimisers/topo_optimiser.rs:
```diff
@@ -123,7 +124,12 @@ where
+        self.run_w_rng(&mut rng)
+    }
+
+    pub fn run_w_rng(mut self, rng: &mut impl RandomSource) -> Result<PhyloOptimisationResult<C>> {
```
This is cool, I assume just for testing purposes? I think overall good implementation with run having the default generator.

junniest replied:
Good catch. This was a copy-paste from the NJ functionality, and this should be handled differently (imo the optimiser should be configured with a random source). I didn't do that in NJ because it would just make any merging more messy, but eventually we should also configure the NJBuilder with a RandomSource.

kalxed commented on phylo/src/optimisers/topo_optimiser.rs:
```diff
@@ -137,14 +143,7 @@ where
-            let rng = &mut rand::thread_rng();
-        }
-        }
+        let mut current_prunes: Vec<_> = possible_prunes.iter().collect();
```
So sauber! Great rework!

kalxed commented on phylo/src/optimisers/topo_optimiser_tests.rs:
```diff
@@ -12,6 +12,7 @@ use crate::optimisers::{
 use crate::parsimony::{scoring::ModelScoringBuilder, BasicParsimonyCost, DolloParsimonyCost};
 use crate::phylo_info::{PhyloInfo, PhyloInfoBuilder as PIB};
 use crate::pip_model::{PIPCost, PIPCostBuilder as PIPCB, PIPModel};
+use crate::random::FakeGenerator;
```
General principle is that fake generator is used in every test file?

junniest replied:
Wherever we need a random generator but need the results to be deterministic, yes!

kalxed commented on phylo/src/random/fake_random.rs:
Very nicely done, tests look good too!

kalxed commented on phylo/src/random/mod.rs:
```diff
@@ -0,0 +1,205 @@
+    }
+
+    #[test]
+    fn different_seeds_produce_different_values() {
```
Just want to highlight this is not necessarily true, though statistically probably will never happen. I think it's a fair test in the way that it's done. Unless implementation of StdRng changes (which is could) it shouldn't fail, and in that case you change the seeding values. Sorry if this is a completely unnecessary comment, as I don't think there is a better way to mitigate, or if it's even necessary.

junniest replied:
I agree with you on this. It is definitely not guaranteed and it is possible that the generated values are the same. Though since the sampled values are f64, I think it is extremely unlikely.
I also can't see a better way to test this, but I definitely think that this test warrants a comment addressing your point.

kalxed commented on phylo/src/random/mod.rs:
```diff
@@ -0,0 +1,205 @@
+    #[test]
+    fn rng_range() {
+        let rng = DefaultGenerator::new(123);
+        for _ in 0..10 {
```
I was wondering for many of the tests whether the amount of times you do it matters? Can't hurt is my thought

junniest replied:
Good question. I have no idea, sadly (:  While googling this the only place I got to is a reddit rabbit hole, so I'm going to keep it as is for now.

kalxed commented on phylo/src/random/mod.rs:
```diff
@@ -0,0 +1,205 @@
+
+impl Default for DefaultGenerator {
+    fn default() -> Self {
+        let seed = Timestamp::now().as_u64(); // Use current timestamp as seed
```
I was curious and this will only be a problem after something like 600k years, how nice.

junniest replied:
The only potential issue of using the timestamp as the seed is if two independent processes are started within the same microsecond, which would produce the same seed and so the same string of pseudo random numbers. But I find the risk of that acceptable :D

kalxed commented on phylo/src/tree/mod.rs:
```diff
@@ -411,28 +412,20 @@ fn argmin_wo_diagonal(q: Mat) -> (usize, usize) {
-#[cfg(not(feature = "deterministic"))]
-fn rng_len(l: usize) -> usize {
-    rand::random::<usize>() % l
+    arg_min[rng.gen::<usize>() % arg_min.len()]
```
I checked this one for myself, and I got issues when generating a usize just off the rip from StdRng generator, but since there is the downcasting implemented I think it might be fine? Regardless I assume this basically generates a u64 to randomize the index?

junniest replied:
Hm, that's interesting, I can straight up generate usize from StdRng. The downcasting is only for the fake_generator so that we don't have to populate an infinite number of sources.

kalxed reviewed (approved):
Looks great to me, will need to rework my code with this but overall a great implementation, there is a bit lost in the very close details but I meticulously checked the wiki with most of my confusion so I think everything makes sense

junniest committed:
Add tests for FakeGenerator coverage

Added a test that requests an unimplemented type and a gen_bool(p) test.

junniest committed:
RandomGenerator test cleanup

junniest committed:
Add comment to diff seed test about no guarantee

Added a comment to the test with different seeds producing different first
values, since it is not guaranteed but extremely unlikely to fail.

junniest committed:
Add test for generating basic types

Added a test for generating different basic types with RandomSource just to
make sure it does not panic.

junniest committed:
Add .seed() to RandomSource and impl

Add a .seed() method to RandomSource trait and implement it for RandomGenerator and
FakeGenerator, add tests for getting the seed.

junniest committed:
Add char to FakeGenerator, change unsupported test

Added char to FakeGenerator supported types, changed test to crash on isize

junniest committed:
Clean up test code -- remove unnecessary typing

junniest committed:
Add test for DefaultGenerator::default()

Added test for DefaultGenerator::default() which should be initialised
with the current timestamp as the seed.

junniest committed:
Remove unnecessary mut from RandomSource::shuffle

junniest committed:
Rename run_w_rng back to run in topo_optimiser

junniest committed:
Fix topo_optimiser doctest

junniest committed:
Fix tests for CI/CD

junniest committed:
Add comment on current rng choice

junniest committed:
Add new_with_pred tests to topo_optimiser

junniest committed:
Remove tests from coverage

junniest committed:
Add char generation to fake_generator

merlinio2000 commented on phylo/src/random/mod.rs:
```diff
@@ -0,0 +1,283 @@
+pub use fake_random::*;
+
+/// Trait for random number generation
+pub trait RandomSource {
```
I don't quite understand why you have to define a trait containing seed, gen, gen_bool and all that stuff.
Is your goal to be able to swap out  `rand` for another library at will?

Otherwise, I would suggest creating an extension trait only containing your specifically needed stuff like `gen_probablility`

junniest replied:
Yes, the idea was to be able to use whatever rng people want, and also to swap out the default here to something platform-independent.

Unfortunately I didn't get it to work just yet because of some weird dependency clash, both `rand_pcg` and `rand_chacha` crates cause the clash. I think that the clashing dependency comes from `bio`, but I'll have to dig more to see what it is.

The other issue is that the way the standard rngs are seeded is different to what people normally use in comp bio. We tend to use a single number because it's easier to reproduce then and no one cares about the rng being crypto-secure, it just needs to be good enough. And for these kinds of applications I feel like we should be able to get the seed back e.g. for debug, which is not something that the `SeededRng` trait provides (for a good reason, but again, here we just need it to be good enough, not secure). I don't see how it would be possible without defining an additional structure like the `SeededRng` below, but please correct me if I'm wrong! (And that's a bad name for it too, since it's almost the same as `SeedableRng`, same for `RandomSource`...)

The other thing is that I defined a trait for this is so that I can make a fake generator that we can supply to tests. Implementing the `Rng` and `SeedableRng` for a fake was somehow very clunky so I ended up with this instead.

I am very certain that this is not the greatest design, just the best thing I could come up with in a limited time. Very happy to hear any suggestion for improving this.

merlinio2000 replied:
Ah yes that totally makes sense, especially the seeding. Crytpographic random providers are definitely overkill for you. Just using unix/libc `rand` might work, but at the top of my head I don't know how that behaves in a multi threaded context.

The clashes between rand_ libraries and bio seem curious. I can't try it out right now but have you tried to import the rand libraries under an alias? Then there should definitely not be a conflict. In general a single crate can be instantiated multiple times with different version in the same build no problem.

About reproducibility: should this eventually be able to be 'deterministic'? In that case I thinks it unavoidable that you generate one giant sequence of random numbers and then distribute slices of values to threads in a deterministic order.

junniest replied:
Yes, the clash is very strange to be honest. I suspect that `bio` uses rand somewhere and defines something that depends on it, but I am legitimately only using the fasta reader from that crate which shouldn't be causing trouble... And I agree that there shouldn't be a dependency clash :D but for some reason cargo thinks otherwise and I've given up on messing with it for now, I am working on running the benchmarks at the moment. Obviously, our cluster setup is entirely different to what you guys have, so that's just great fun.

For reproducibility: I think that I ended up confusing the determinism in terms of computational reproducibility and the determinism needed for tests. The FakeGenerator is good enough for tests so far, and that's the only true determinism that we need. The version of determinism for computational reproducibility (and that only applies to the jati binary and the same version of that binary) just means that if I run the binary twice with the same seed I should get identical logs (with the exception of timestamps in the log). The good thing is that it actually seems to be working as intended with the current implementation, but I've only tested it manually so far (i.e. ran the thing twice with the seed of 1 and got the exact same numbers). I will add tests that verify this to the binary, but the library should not care about that, I think.

And yes, if we ever need the threads to also use something randomised, what you say seems like a good idea.

merlinio2000 commented on phylo/src/random/mod.rs:
```diff
@@ -0,0 +1,283 @@
+where
+    R: Rng + SeedableRng + Send,
+{
+    r: Mutex<SeededRng<R>>,
```
No me gusta, since the order when multithreading is out the window anyway (no reproducibility) I think it would make much more sense to work with some sort of threadlocal storage

junniest replied:
Yes, totally agree. I was already thinking about that and at the time made the executive decision to not overthink it because a student was blocked by needing the `FakeRandom`.

The other thought here though is that we are not obliged to provide a perfectly reproducible version of a random generator that will handle threads in a reproducible way too. I am also not entirely sure it is possible to enforce either. 

In any case, at the moment the parallel parts of the spr move evaluation don't use any rng, rng is used to shuffle the pruning locations and then the threads work deterministically. Of course, if they ever need rng, this will be a different story, but I think I have a way in mind how to keep the reproducibility in the JATI binary without sacrificing parallel computing (supplying local randomisers seeded by values from the overall randomiser). 

Perhaps I should in general rename and comment on these structs to state clearly that this is not a general purpose random provider, but a very specifically limited version meant to promote computational reproducibility in phylogenetics. I still don't think that we can enforce completely deterministic behaviour on the crate users in general, so it might just be their individual responsibility to make sure their code is reproducible in the end.

Happy to hear your thoughts on this. This turned out surprisingly difficult to design in a meaningful way, and most likely should be refactored asap anyway (:

merlinio2000 replied:
I keep forgetting that people will also use this as a library not just a binary, thanks. Renaming this to be clear that its not meant for general purpose use (and maybe also make it private) seems like a good idea.

From a library design point, passing the generator explicitly around and not making it some global state is probably the most important part. Another thing about library usage, your RNG trait currently only takes immutable references to self, this is only made possible because Mutex allows interior mutability. I suggest making the trait take self by `&mut` (which `rand_` also does) as the generator carries an internal state that gets modified for each new number. If you need the shared references (e.g. passing to multiple threads) I would make that the users worry to properly solve as hes probably in dangerous territory anyways

junniest replied:
Yes, I will refactor this as soon as I have the time. Good call on the mutability, since the `Mutex` in any case will not make this reproducible if a bunch of threads also ask for randomisation. And I agree, I get mixed up in the library/binary specific requirements, but the binary requirements that I personally have should not apply to the library, it's the user's concern.

Unfortunately (or fortunately?), I specifically made a separate repo for the jati binary to also make sure that the library works as intended from an external crate. But this also means that the random generators cannot be private, since then I won't be able to use them either :D But I can use them as intended (and that way maybe also provide a usage example) and make sure that the same seeds generate the same results. As I mentioned in the other comment, on first glance it seems like it, and I think that's good enough for me.
